lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -1719,16 +1771,59 @@ class SemanticsObject {
     }
   }
 
-  /// Recursively visits the tree rooted at `this` node in depth-first fashion.
+  /// Recursively visits the tree rooted at `this` node in depth-first fashion
+  /// in render order.
   ///
   /// Calls the [callback] for `this` node, then for all of its descendants.
-  void visitDepthFirst(void Function(SemanticsObject) callback) {
+  void visitDepthFirstInRenderOrder(void Function(SemanticsObject) callback) {
     callback(this);
     _currentChildrenInRenderOrder?.forEach((SemanticsObject child) {
-      child.visitDepthFirst(callback);
+      child.visitDepthFirstInRenderOrder(callback);
```

Comment:
I looked at the code the `_currentChildrenInRenderOrder` is based on `childrenInTraversalOrder`, what is the difference between this and `visitDepthFirstInTraversalOrder`

Reply:
`childrenInRenderOrder` is the order the child elements are rendered into the DOM. IOW, if you do `containerElement.elements` you will get children listed in `childrenInRenderOrder`. Separately each child may have a `z-index`, which dictates its hit test order relative to siblings (also paint order, but our semantics DOM is invisible so that doesn't matter).

Effectively, `childrenInRenderOrder` dictates the traversal order in the DOM. This is the order screen readers will observe. The logic that translates `childrenInTraversalOrder`/`childrenInHitTestOrder` to `childrenInRenderOrder`/`z-index` is in `updateChildren()`. We use a different name because it does not guarantee every possible traversal order the framework could use. The web limitations are similar to iOS though, so I think in practice they are the same.

The reason this function uses `childrenInRenderOrder` is to be in sync with the order that the screen reader observes.

Reply:
I understand the difference between childrenInHitTestOrder and visitDepthFirstInRenderOrder. I was confused at the difference between `visitDepthFirstInRenderOrder` and `visitDepthFirstInTraversalOrder`, they seem to be the same.

Reply:
Ah, I see. You are correct. If the tree is fully consistent the two should effectively return the same result. Currently, the only useful difference is that `visitDepthFirstInTraversalOrder` ensures that the tree is finalized and is in consistent state before you traverse it. There are some asserts that enforce it. However, `visitDepthFirstInRenderOrder` can operate on unfinalized/inconsistent tree; it just returns partial results, which is useful for debugging and assertion messages. For example, `_computeNodeMapConsistencyMessage` uses it and it generates an error that fully explains where things are amiss. If it tried using `visitDepthFirstInTraversalOrder` it wouldn't be able to print the partial message because `visitDepthFirstInTraversalOrder` would crash on any missing node in the tree.

However, this different doesn't seem to be too interesting. Let me see if I can clean it up.

Reply:
I couldn't find a way to fully remove `visitDepthFirstInRenderOrder` because it's still needed to be able to walk a potentially broken tree (for `assert`s). But it's a private method only, so I made it private and debug-only, added a doc about its difference from the other method, and stopped using it in a non-debug context.